### PR TITLE
Added a regex format for date for use in customer

### PR DESCRIPTION
**Pull request by cneijenhuis**
Found this bug thanks to @nkuehn validator :)

The regex should be correct for values we emit.

For create/update, our parser actually accepts more values than the regex does... If anyone wants to, they can build a regex from this: https://github.com/JodaOrg/joda-time/blob/a3b1a16ee43af611939799a99c7c461d051d40e7/src/main/java/org/joda/time/format/ISODateTimeFormat.java#L1345-L1370 :wink: However, I don't really see a need to support formats other than YYYY-MM-DD (including, e.g. YYYY).

Please review @agourlay or @OlegIlyenko

### CODE REVIEW

File: schemas/date.schema.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "$schema":"http://json-schema.org/draft-04/schema#",
+
+  "type":"string",
+  "pattern":"^\d{4}-((0[1-9])|(1[012]))-((0[1-9]|[12]\d)|3[01])$"
```
Comment (hajoeichler) [inline schemas/date.schema.json]: Do we need all those brackets here?
what about pretty simple: `^\d{4}-[0-1]\d-[0-3]\d$`

Comment (nkuehn) [inline schemas/date.schema.json]: I also don't think we need the capture groups. They would just be necessary for parsing the String via regexp, but we just want to validate.

**Comment by cneijenhuis** [comment]
Oleg pointed me into the right direction. Actually all formats defined in the RFC can be used in json schema, not only date-time.

**Comment by nkuehn** [comment]
FYI:   I have just "travisified" my branch.  
I merged this change into my branch but the test still fails ( https://travis-ci.org/sphereio/sphere-api-reference/jobs/68797074#L138 ). The metaschema doesn't like "full-date" either or something else.  
Before merging my branch we'll have to find out whether the way I validate is insufficient or what else is the issue here.

**Comment by cneijenhuis** [comment]
Hmrpf... the spec says:

> date-time
> A string instance is valid against this attribute if it is a valid date representation as defined by RFC 3339, section 5.6 [RFC3339]. 

Only that section 5.6 defines several date (and time) representations. So it's unclear whether they refer to all representations or only to `date-time`.

After reading this comment: https://github.com/natesilva/jayschema/issues/19 where he says that the json schema v3 also defined `date` and `time`, but removed both in v4, I think my initial understanding of it was right (= they only support `date-time`).

File: schemas/date.schema.json (hunk 1 above)
Comment (cneijenhuis) [inline schemas/date.schema.json]: @hajoeichler because 2015-00-32 would be valid for your regex, but not in the other one.
Actually I'm also fine with `d{4}-d{2}-d{2}` - it's much quicker to understand and all sorts of invalid dates will not be able to be catched with a regex anyway (leap years and what not :wink:)

**Comment by OlegIlyenko** [comment]
hmmm... it's a pity.

**Comment by nkuehn** [comment]
Hi everybody, could it be that the actual issue why the stuff does not validate against the meta-schema is something else:
instead of 
```json
    "dateOfBirth": {
      "type":"full-date"
    },
```
it should be 
```json
    "dateOfBirth": {
      "type":"String",
      "format":"full-date"
    },
```
Formally, my validator finds this OK. 

It's possible to define custom formats, too if I got that right from the spec.

**Comment by cneijenhuis** [comment]
Argh, you are right - yes.

I think the validator can define/support additional formats. I can't really see in the spec what the validator should do if it encounters an unknowm format. Maybe it just ignores it? It would be good to test whether the validator will actually only accept YYYY-MM-DD or any string.

**Comment by nkuehn** [comment]
check it out and try what the validator does then!   

I wouldn't work with custom formats but rather with regular expressions because custom formats make the spec dependent on validator and tool support and that's a bad thing for a public spec that's meant to be friendly to externals.

**Comment by cneijenhuis** [comment]
I finally found some time to try the validator myself @nkuehn :)
Anyway, `"format":"full-date"` passes through the schema-validator, but the example validity check says:

**could not instantiate validator for schema: Project (/{projectKey}) -> get -> responses -> HTTP 200 -> body -> application/json**
```
[Error: Invalid format_name for 'format']
```

So, I guess we can agree now that we won't use `full-date` :wink: Since you didn't like the complex regexp, I've now used a simple one. I have confirmed with the example validator that it's actually being used to check the dateOfBirth.

Can I merge this now? @nkuehn @hajoeichler

**Comment by hajoeichler** [comment]
:+1:

**Comment by nkuehn** [comment]
Christoph, here you can find the format names supported by the draft 4
validator my script uses:
https://github.com/pandastrike/jsck/blob/1c47c4a3579db2bf3882120f45e05702418ec148/src/draft4/strings.coffee#L45


The list goes even further than what I can find in the JSON schema draft 4
RFC:
http://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-7 .
If I get the wording right, the JSON schema spec supports only the
"date-time" variant of  RFC3339 (time formats).

The JSCK validator regexpes were copied from another validator
implementation, which seems to have a relatively loose approach to date /
time / date-time (supports the draft 3 variants, too).

TL;DR: +1 ,  there is no explicit "full-date" in draft4 JSON schema
anyways.

2015-07-06 21:15 GMT+02:00 Hajo <notifications@github.com>:

> [image: :+1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sphereio/sphere-api-reference/pull/3#issuecomment-118967893>
> .
>



-- 
Nikolaus Kühn
Partner Enablement Manager

[image: commercetools logo] <http://www.commercetools.de/>

E. nikolaus.kuehn@commercetools.de
M. +49 172 2873 155 <+491722873155>
T. @nikolausku <https://twitter.com/nikolausku>
Adams-Lehmann-Str. 44, 80797 München
commercetools.de <http://www.commercetools.de/>

-- 


<http://ecommerceopen.de/>

commercetools GmbH
SPHERE.IO <http://www.sphere.io/> - The E-Commerce Platform.

Amtsgericht München, HRB 161496
Geschäftsführer: Denis Werner, Dirk Hörig, Dr. Robert Zores

www.commercetools.com <http://www.sphere.io/>

**Comment by cneijenhuis** [comment]
Thanks for the  link to the regex-es. They have implemented `date` (semantically the same as `full-date`), which was supported in draft 3 but removed in draft 4.
Let's review on a new draft, if support for additional formats is added... I'll merge this now. :+1:
